From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Wed, 29 Jul 2026 02:30:35 +0200
Subject: media: ov5640: Add brightness control

struct ov5640_ctrls has had a brightness member since forever, but the
control was never created. The sensor applies it as the Y bright
offset of the contrast function, a magnitude in SDE CTRL7 with its
sign in SDE CTRL8.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/media/i2c/ov5640.c | 31 ++++++++++
 1 file changed, 31 insertions(+)

diff --git a/drivers/media/i2c/ov5640.c b/drivers/media/i2c/ov5640.c
index 111111111111..222222222222 100644
--- a/drivers/media/i2c/ov5640.c
+++ b/drivers/media/i2c/ov5640.c
@@ -129,6 +129,7 @@
 #define OV5640_REG_SDE_CTRL4		0x5584
 #define OV5640_REG_SDE_CTRL5		0x5585
 #define OV5640_REG_SDE_CTRL6		0x5586
+#define OV5640_REG_SDE_CTRL7		0x5587
 #define OV5640_SDE_CTRL0_SAT_EN		BIT(1)
 #define OV5640_SDE_CTRL0_CONTRAST_EN	BIT(2)
 /* UV adjust gain, 0x40 is unity */
@@ -147,6 +148,7 @@
  */
 #define OV5640_SDE_CTRL8_HUE_SIN_POS	BIT(0)
 #define OV5640_SDE_CTRL8_HUE_SIN_NEG	BIT(1)
+#define OV5640_SDE_CTRL8_BRIGHT_NEG	BIT(3)
 #define OV5640_SDE_CTRL8_HUE_COS_NEG	(BIT(4) | BIT(5))
 #define OV5640_SDE_CTRL8_HUE_SIGN_MASK	(OV5640_SDE_CTRL8_HUE_SIN_POS | \
 					 OV5640_SDE_CTRL8_HUE_SIN_NEG | \
@@ -3486,6 +3488,30 @@ static int ov5640_set_ctrl_contrast(struct ov5640_dev *sensor, int value)
 	return ov5640_write_reg(sensor, OV5640_REG_SDE_CTRL6, value);
 }
 
+static int ov5640_set_ctrl_brightness(struct ov5640_dev *sensor, int value)
+{
+	int ret;
+
+	/*
+	 * The brightness is the Y bright offset in SDE CTRL7, a magnitude
+	 * with its sign in SDE CTRL8. It is part of the contrast function,
+	 * so that has to be enabled for it to take effect.
+	 */
+	ret = ov5640_mod_reg(sensor, OV5640_REG_SDE_CTRL0,
+			     OV5640_SDE_CTRL0_CONTRAST_EN,
+			     OV5640_SDE_CTRL0_CONTRAST_EN);
+	if (ret)
+		return ret;
+
+	ret = ov5640_mod_reg(sensor, OV5640_REG_SDE_CTRL8,
+			     OV5640_SDE_CTRL8_BRIGHT_NEG,
+			     value < 0 ? OV5640_SDE_CTRL8_BRIGHT_NEG : 0);
+	if (ret)
+		return ret;
+
+	return ov5640_write_reg(sensor, OV5640_REG_SDE_CTRL7, abs(value));
+}
+
 static int ov5640_set_ctrl_saturation(struct ov5640_dev *sensor, int value)
 {
 	int ret;
@@ -3839,6 +3865,9 @@ static int ov5640_s_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_HUE:
 		ret = ov5640_set_ctrl_hue(sensor, ctrl->val);
 		break;
+	case V4L2_CID_BRIGHTNESS:
+		ret = ov5640_set_ctrl_brightness(sensor, ctrl->val);
+		break;
 	case V4L2_CID_CONTRAST:
 		ret = ov5640_set_ctrl_contrast(sensor, ctrl->val);
 		break;
@@ -3958,6 +3987,8 @@ static int ov5640_init_controls(struct ov5640_dev *sensor)
 					      OV5640_SATURATION_UNITY);
 	ctrls->hue = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_HUE,
 				       0, 359, 1, 0);
+	ctrls->brightness = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_BRIGHTNESS,
+					      -255, 255, 1, 0);
 	ctrls->contrast = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_CONTRAST,
 					    0, 255, 1, OV5640_CONTRAST_UNITY);
 	ctrls->test_pattern =
-- 
Armbian

